ydb/core/tx/datashard/prefix_kmeans.cpp

```diff
@@ -356,20 +354,38 @@ class TPrefixKMeansScanBase: public TActor<TPrefixKMeansScanBase>, public NTable
         UploadImpl();
     }
 
+    bool UploadInProgress()
+    {
+        return !UploadBuf.IsEmpty();
+    }
+
+    bool TryUpload(TBufferData& buffer, const TString& table, const std::shared_ptr<NTxProxy::TUploadTypes>& types, bool byLimit)
+    {
+        if (Y_UNLIKELY(UploadInProgress())) {
+            // already uploading something
+            return true;
+        }
+
+        if (!buffer.IsEmpty() && (!byLimit || buffer.IsReachLimits(Limits))) {
+            buffer.FlushTo(UploadBuf);
+            InitUpload(table, types);
+            return true;
+        }
+
+        return false;
+    }
 
-    void UploadSample()
+    void WriteLevelRows()
```

Comment:
UploadLevelRows?

Reply:
Tried to separate 'upload' as writing on datashards with these 'form rows and save them to buffer' behaviours